hal/src/rtl872x/i2c_hal.cpp

```diff
@@ -502,58 +578,140 @@ class I2cClass {
         return false;
     }
 
-    void slaveReadRxFifo() {
-        while (I2C_CheckFlagState(i2cDev_, (BIT_IC_STATUS_RFNE | BIT_IC_STATUS_RFF))) {
-            rxBuffer_.put((uint8_t)I2C_ReceiveData(i2cDev_));
+    bool handleStartStop() {
+        // These bits may be both set in single ISR
+        // start | stop
+        //   0   |   1    : STOPPED
+        //   1   |   0    : RESTARTED
+        //   1   |   1    : STOPPED -> STARTED (most likely happen) or RESTARTED -> STOPPED
+        uint32_t intStatus = I2C_GetINT(i2cDev_);
+        if ((intStatus & BIT_IC_INTR_STAT_R_START_DET) || (intStatus & BIT_IC_INTR_STAT_R_STOP_DET)) {
+            I2C_ClearINT(i2cDev_, BIT_IC_INTR_STAT_R_START_DET);
+            I2C_ClearINT(i2cDev_, BIT_IC_INTR_STAT_R_STOP_DET);
+            if ((intStatus & BIT_IC_INTR_STAT_R_START_DET) && !(intStatus & BIT_IC_INTR_STAT_R_STOP_DET)) {
+                slaveStatus_ = I2C_SLAVE_RESTARTED;
+            } else if (intStatus & BIT_IC_INTR_STAT_R_START_DET) {
+                slaveStatus_ = I2C_SLAVE_STARTED;
+            } else {
+                // FIXME: It might be RESTARTED -> STOPPED
+                slaveStatus_ = I2C_SLAVE_STOPPED;
+            }
+            return true;
         }
-        if (onReceived_ && rxBuffer_.data() > 0) {
+        return false;
+    }
+
+    void slaveRxDone(size_t len, bool reset = false) {
+        rxBuffer_.put(slaveRxCache_.get(), len);
+        if (onReceived_) {
             onReceived_(rxBuffer_.data());
         }
+        rxBuffer_.reset();
+        if (reset) {
+            memset(slaveRxCache_.get(), 0xFF, SLAVE_RX_CACHE_DEPTH);
+            slaveRxCacheLen_ = 0;
+        } else {
+            slaveRxCacheLen_ = slaveRxCacheLen_ - len;
+            std::memmove(slaveRxCache_.get(), &slaveRxCache_[len], slaveRxCacheLen_);
+        }
     }
 
-    static void i2cSlaveIntHandler(void* context) {
-        auto instance = (I2cClass*)context;
-        uint32_t intStatus = I2C_GetINT(instance->i2cDev_);
-
-        if (intStatus & BIT_IC_INTR_STAT_R_START_DET) {
-            I2C_ClearINT(instance->i2cDev_, BIT_IC_INTR_STAT_R_START_DET);
-            // Read RX FIFO in case of repeated write
-            instance->slaveReadRxFifo();
-            instance->rxBuffer_.reset();
-            instance->txBuffer_.reset();
+    // Return:
+    // 1: some of the read data belongs to next session
+    // 0: all of the read data belongs the current session
+    bool slaveRead() {
+        uint16_t reportLen = 0;
+        uint32_t intStatus;
+        while (I2C_CheckFlagState(i2cDev_, (BIT_IC_STATUS_RFNE | BIT_IC_STATUS_RFF))) {
+            slaveRxCache_[slaveRxCacheLen_++] = (uint8_t)I2C_ReceiveData(i2cDev_);
```

Comment:
@XuGuohui What if `SLAVE_RX_CACHE_DEPTH` is less than RX buffer and the transaction is also over `SLAVE_RX_CACHE_DEPTH` ? We can overflow here looks like.

Why not use the RX buffer itself and extend it over the configured size if required for peripheral behavior weirdness? E.g. x2?